### ISSUE TITLE
🦠 Model Request: Tox21-ER

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#457]
### Model Title

Tox21-ER (TDCommons)

### Publication

Hello @AnneO1!

As part of your Outreachy contribution, we have assigned you the dataset "Tox21-ER" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by AnneO1** [comment]
Thanks at @GemmaTuron, on it.

**Comment by AnneO1** [comment]
Hi @GemmaTuron,
Tox21 contains multiple assays data and to retrieve the specific labels for an assay, one specifies the label name. When I checked the label name list, I could not find 'Tox21-ER' but I saw 'Tox21 NR-ER'. Kindly assist since am a little confused.

Notebook : https://colab.research.google.com/drive/1mkHWLq0QqWx_ma_zFkwuXgpx9_6Nwz6u?usp=sharing



![Capture error](https://user-images.githubusercontent.com/115647922/198642082-dbf98d63-42c6-45a0-8d61-df475d22101c.PNG)

**Comment by GemmaTuron** [comment]
Hello @AnneO1 !

I can't access your google colab
Apologies about the confusion, please use the NR-ER dataset

**Comment by AnneO1** [comment]
Thanks Gemma, I will finalize and add the Colab link.

**Comment by AnneO1** [comment]
Hello @GemmaTuron,

I have been trying to finalize with training model Tox21 NR-ER. Here is a link to my notebook:
https://colab.research.google.com/drive/1mkHWLq0QqWx_ma_zFkwuXgpx9_6Nwz6u?usp=sharing

**Comment by AnneO1** [comment]
**Problem:**
Getting Tox21 NR-ER from the Therapeutics Data Commons and build a binary classification ML model.
**Model Info:**
Tox21 is a data challenge which contains qualitative toxicity measurements for 7,831 compounds. It contains multiple assays data and Tox21 NR-ER (Nuclear Receptor- Estrogen Receptor ) is a label name for the assay used in this modeling.

**Data Loading**
I installed pyTDC the Therapeutics Data Commons(TDC) package in order to download the Tox21 dataset. I then loaded the NR-ER data from the available label names present in Tox21 dataset as seen below:
![1](https://user-images.githubusercontent.com/115647922/199493388-bbfabb80-f014-421b-986d-ee5b98233576.PNG)

**Data Preparing**
A key part in model-building project is separating the data into:
Train: the portion of the dataset that will be used to train the model.
Valid: the validation set, which will be used during the modelling to assert and improve the performance of the algorithm.
Test: a portion of the dataset kept completely separated from the data and will be used to evaluate the final model performance.
I then used Scaffold method to split the data. Scaffold split is based on the scaffold of the molecules so that train/val/test set is more structurally different and only applies to single-instance drugs like Tox. I then used Pandas head() method to get an overview of the data:
![image](https://user-images.githubusercontent.com/115647922/199495054-743565c8-7d8a-473c-80f3-1766ee27a6f7.png)
Column Drug represents the molecules in Simplified Molecular Input Line Entry System (SMILES), a chemical notation for chemical structure that allows the computer to understand. Column Y represents values 0 or 1 a classification for active(toxic) 1 or inactive(non-toxic) 1 useful as this is a Binary Classification Model.

**Analysis**
_Prediction?_
For dataset Tox21 NR-ER, I am trying to predict molecule toxicity. From column Y in the dataset, it is possible to see the classification of the molecules and their effects on humans presented as either 0 or 1.

_Is this a classification or a regression problem?_ 
This is a Binary Classification Model since data is presented in either 0 or 1 and as such a Classification Problem.

_How many data points in total and in each set (train / validation / test)?_
The following is the total number of molecules available in the three sections:
![image](https://user-images.githubusercontent.com/115647922/199500461-00958bf2-28db-4127-8530-95e562ace52c.png)

_How many actives and inactives present?_ 
Per the three sections, the active and inactive are as follows:
![image](https://user-images.githubusercontent.com/115647922/199501179-5d03c287-e51f-4873-8a61-30475a27d05b.png)

In order to explore the chemical structures of the data, leverage an open source toolbox for cheminformatics RDKit to visualize the molecules.
_Active Molecules:_
![image](https://user-images.githubusercontent.com/115647922/199507215-ede83ba7-c811-4ccc-bf19-39cedf4e7263.png)

_Inactive Molecules:_
![image](https://user-images.githubusercontent.com/115647922/199507592-4b8d5fda-badd-4f77-a219-5138caff0dbb.png)

The distribution per the interested section(train dataset)
![image](https://user-images.githubusercontent.com/115647922/199505989-84de3d1d-ddca-477e-83b4-577fe9805209.png)
**In general, we can derive that in each of the datasets, the inactive molecules are greater than the active molecules therefore an unbalanced dataset.**

**Comment by AnneO1** [comment]
**Model Training**
I used LazyQsar package from Ersilia to prepare my data for modelling since it automatically converts smiles to Morgan fingerprints and is very efficient. I then imported the Morgan Binary Classifier class, instantiated it and used the fit function for training. Fitting the model to the data: the molecules and outputs of the Train set.
Time taken to find the best model: 6.70 and saved the model in joblib format for faster loading.

Model Evaluation
We can use the following measures to understand whether the model is performing correctly or not:
ROC Curve: a probability curve showing the True Positive and False Positive Rates at different classification thresholds.
To test the results of our prediction, I used a different dataset not the train set to evaluate the performance of the model. I got predictions for the validation and test dataset as probabilities. Since this is a Classification problem, it's more flexible to predict the probabilities for each class which provides the capability to choose and even calibrate the threshold for how to interpret the predicted probabilities.
![image](https://user-images.githubusercontent.com/115647922/199582421-f07ed814-641d-4ecf-b78f-8330501ba56d.png)
The AUC value is 0.70644 which is a good indication.
Confusion matrix: a table that indicates how many molecules were correctly classified by the model and how many were misclassified. I used a Sckit learn python library to get the confusion matrix for the validation set as seen below:
![image](https://user-images.githubusercontent.com/115647922/199583357-c1b85a37-bc1d-43f3-9eb9-55a80a4b790c.png)
The outcomes of the matrix was the following
True Positive: (3) model correctly predicts the positive class meaning our model predicts a drug as inactive(non-toxic) and its a correct prediction aka hit
True Negative: (516) model correctly predicts as non toxic and were inactive thus true negative.
False Positive: (0)  model incorrectly predicts as toxic, when they were negative, non-toxic.
False Negative: (100) model predicted them as non-toxic when infact they are toxic thus a miss.

Precision: number of true positives divided by the number of true positives plus the number of false positives. It reflects how reliable the model is in classifying samples as Positive.
Recall: The number of instances which the model correctly identified as relevant out of the total relevant instances.
![image](https://user-images.githubusercontent.com/115647922/199595087-ba09480e-e7ff-42bf-9c52-7a9f8eabbef2.png)

**Comment by GemmaTuron** [comment]
Hello @AnneO1  !

Thanks for your contributions to Ersilia during the Outreachy Application Period! We hope you have learnt and enjoyed as much as the Ersilia team did!
I will close this issue as this was part of the application period solely.

**Comment by AnneO1** [comment]
Hi Gemma, I really did. Thank you so much for having us and for taking us through this journey. It was a huge step for me. Thanks again. Happy times ahead.
